modules/data.remote/R/remote_process.R

```diff
@@ -0,0 +1,703 @@
+##' @examples
+##' \dontrun{
+##' remote_process(settings)
+##' }
+##' @author Ayush Prasad
+##'
+
+remote_process <- function(settings) {
+  # information about the date variables used in remote_process -
+  # req_start, req_end : start, end dates requested by the user, the user does not have to be aware about the status of the requested file in the DB
+  # start, end : effective start, end dates created after checking the DB status. These dates are sent to remote_process for downloading and processing data
+  # write_raw_start, write_raw_end : start, end dates which are used while inserting and updating the DB
+  # the "pro" version of these variables have the same meaning and are used to refer to the processed file
+  
+  RpTools <- reticulate::import("RpTools")
+  
+  input_file <- NULL
+  stage_get_data <- NULL
+  stage_process_data <- NULL
+  raw_merge <- NULL
+  pro_merge <- NULL
+  existing_raw_file_path <- NULL
+  existing_pro_file_path <- NULL
+  
+  # extract the variables from the settings list
+  siteid <- as.numeric(settings$run$site$id)
+  siteid_short <- paste0(siteid %/% 1e+09, "-", siteid %% 1e+09)
+  raw_mimetype <- settings$remotedata$raw_mimetype
+  raw_formatname <- settings$remotedata$raw_formatname
+  outdir <- settings$outdir
+  start <- as.character(as.Date(settings$run$start.date))
+  end <- as.character(as.Date(settings$run$end.date))
+  source <- settings$remotedata$source
+  collection <- settings$remotedata$collection
+  scale <- settings$remotedata$scale
+  if (!is.null(scale)) {
+    scale <- as.double(settings$remotedata$scale)
+    scale <- format(scale, nsmall = 1)
+  }
+  projection <- settings$remotedata$projection
+  qc <- settings$remotedata$qc
+  if (!is.null(qc)) {
+    qc <- as.double(settings$remotedata$qc)
+    qc <- format(qc, nsmall = 1)
+  }
+  algorithm <- settings$remotedata$algorithm
+  credfile <- settings$remotedata$credfile
+  pro_mimetype <- settings$remotedata$pro_mimetype
+  pro_formatname <- settings$remotedata$pro_formatname
+  out_get_data <- settings$remotedata$out_get_data
+  out_process_data <- settings$remotedata$out_process_data
+  
+  
+  dbcon <- PEcAn.DB::db.open(settings$database$bety)
+  on.exit(PEcAn.DB::db.close(dbcon), add = TRUE)
+  
+  flag <- 0
+  
+  # collection dataframe used to map Google Earth Engine collection names to their PEcAn specific names
+  collection_lut <- data.frame(
+    stringsAsFactors = FALSE,
+    original_name = c(
+      "LANDSAT/LC08/C01/T1_SR",
+      "COPERNICUS/S2_SR",
+      "NASA_USDA/HSL/SMAP_soil_moisture"
+    ),
+    pecan_code = c("l8", "s2", "smap")
+  )
+  getpecancode <- collection_lut$pecan_code
+  names(getpecancode) <- collection_lut$original_name
+  
+  if (source == "gee") {
+    collection = unname(getpecancode[collection])
+  }
+  
+  req_start <- start
+  req_end <- end
+  
+  raw_file_name = construct_raw_filename(collection, siteid_short, scale, projection, qc)
```

Comment:
If these are the arguments used in the filename, we need to make sure they are not NULL, for example the check above should include an else to use some default (with a message):
```

  if (!is.null(scale)) {
    scale <- as.double(settings$remotedata$scale)
    scale <- format(scale, nsmall = 1)
  }else{
     scale <- 10
     PEcAn.logger::info("No sclae provided, Using 10 as default")
  }
```

Ideally, all remotedata tags should be checked to see if they are filled with sensible values in the beginning, probably you have checks later for some of those later on but I think beginning of the module is a good place to do that.

We should check even if the source is valid, e.g. "GEE" but not "GoogleEarthEngine". If we use registration files we can look if `register.SOURCE.xml` file exists, if not this also means that source is not implemented yet. Overall, I think registration files will help us on many levels

Speaking of which, maybe source should also be in the filename, a filename like `s2_10.0_***_site_1-26929` felt a bit uninformative as opposed to `GEE_s2_10.0_***_site_1-26929` even though the folder name has the source

Reply:
> `     PEcAn.logger::info("No sclae provided, Using 10 as default")`

Currently, all of the functions which require such arguments don't have a default value(except l8 qc which I see I forgot to remove) Do you think they should have a default value? If yes, I think that should not be set here in `remote_process` and should be set in the Python function because the values differ for every collection.

> If these are the arguments used in the filename, we need to make sure they are not NULL

Can you please explain about this? It is being checked if they are NULL and in that case "NA" is used in place of the value in the filename. I went this way so that there would be a same logic for constructing the file name regardless of source and collection.

> Ideally, all remotedata tags should be checked to see if they are filled with sensible values in the beginning

Agreed, I should have done that.

> Speaking of which, maybe source should also be in the filename, a filename like s2_10.0_***_site_1-26929 felt a bit uninformative as opposed to GEE_s2_10.0_***_site_1-26929 even though the folder name has the source

Yes, will change it back.

Reply:
> Currently, all of the functions which require such arguments don't have a default value set in the Python function(except l8 qc which I see I forgot to remove) Do you think they should have a default value? 

Not sure if they have defaults, probably yes because the person who implements/develops the function typically knows better than most of the end users. But even if they don't have a default value, I think the module should fail here when some argument of the downstream function is missing or passed an invalid value.

>If yes, I think that should not be set here in `remote_process` and should be set in the Python function because the values differ for every collection.

I think we can use the registration files for inserting the defaults as well, which solves values differ for every collection and reduce user errors

>  It is being checked if they are NULL and in that case "NA" is used in place of the value in the filename. I went this way so that there would be a same logic for constructing the file name regardless of source and collection.

It is a bit unusual to have `NA` in the file names (maybe even not advisable?). Again maybe registry files can help here, when that xml is read the filename can be constructed from certain tags. Do you have downstream functions that absolutely rely on this filename convention to be preserved regardless of source and collection?

Reply:
I too haven't seen NA being used in filenames. I also used this so that the filenames could be more sensible, example: in GEE_s2_10.0, 10 means scale and in some other collection it might be the qc

>  Do you have downstream functions that absolutely rely on this filename convention to be preserved regardless of source and collection?

Nothing as such, this was mostly to have a uniform naming convention, otherwise, it could be a little confusing for different sources and collections. e.g. appeears has projection but gee doesn't.

 But yeah maybe after trying with registry files I would be able to think better.

Reply:
By the way, this was something we discussed but haven't decided: Currently you keep merging multiple years into single files, is this correct? @mdietze load_data functions can subset files downstream (like the Ameriflux csv files), but following the met example do we want files returned by the remotedata module to be yearly? I.e. are we OK if `GEE_***_site_1-26929` file has multiple years or do we want them split, e.g. `GEE_***_site_1-26929.2018.nc`, `GEE_***_site_1-26929.2019.nc` ...?

Also @mdietze some sources (e.g. AppEEARS) already returns csv format iirc, do we want remote module to reformat them into netcdf?

Reply:
> currently you keep merging multiple years into single files, is this correct?

Yes, they are merged in a single file. This entire module is designed with the idea that there should always be **only 1 file** of the same type in the DB regardless of years.